### ISSUE TITLE
Building libvips as a native gem extension?

### ISSUE THREAD

**Issue by Benjamin-Dobell** [jcupitt/ruby-vips#182]
My apologies if this has already been discussed.

I see at one point there was a native extension being built as part of this gem and that it has since been removed. I'm just wondering what the rational is for requiring manual installation of `libvips`, rather than building `libvips` as a native extension.

There may well be something I'm not considering, however based on the information I have, I would prefer it if `libvips` were built as an extension.

My rationale is that `ruby-vips` is tied to specific `libvips` versions (at the time of writing, 8.6+). If we update our ruby gems, our app may run fine in one environment (i.e. our development machine), but not in another environment (production). Effectively our Ruby apps now have a dependency that isn't declared anywhere, and hence not resolvable using normal dependency management.

Building `libvips` from source may seem like a pain, as it's conceivable that `libvips` has already been installed from a system's package manager. However, realistically any version of `libvips` installed from anywhere other than source is going to be too old for `ruby-vips` anyway. This is the case because production-grade distros don't keep up with `libvips` releases so `apt-get update; apt-get install -y libvips libvips-dev` is unlikely to work. So we end up building from source regardless; except we do so in a non-standard fashion, rather than automating this through our dependency management system (RubyGems).

**Comment by jcupitt** [comment]
Hello @Benjamin-Dobell,

There was a ruby-vips a while ago which was written in C and tied to a specific libvips version, but it was a lot of work to maintain. It was more than 10,000 lines of C, as I recall! And of course you still had to build and install libvips before you could install the extension. 


The current ruby-vips is pure Ruby (so it works on any Ruby platform), only needs the libvips shared library, and should automatically adapt to any libvips version. I think anything after libvips 8.2 is OK. The hope is obviously that decoupling ruby-vips from libvips will reduce maintenance.

You're right that there's an extra level of versioning: you need to think about both the ruby-vips version, and the version of libvips it is picking up. 

ruby-vips does download and build libvips has part of Travis integration, did you see?

https://github.com/jcupitt/ruby-vips/blob/master/.travis.yml

But of course that will only work on Debian-style systems. I think it would be very hard to make it support a wide range of platforms. 

I've been deploying with Docker myself, I don't know if that's an option. I only do very small-scale stuff though, it might well not be appropriate for you.

https://github.com/jcupitt/docker-builds

**Comment by jcupitt** [comment]
(sorry, pressed comment accidentally before I was done, I've edited the answer)

**Comment by Benjamin-Dobell** [comment]
> ruby-vips-0.3-stable is written in C, but it was a lot of work to maintain. It was more than 10,000 lines of code

Ah, yes, having a separate native implementation within `ruby-vips` would indeed be a lot of work. I meant simply building the existing `libvips` as part of this gem. Kind of like what I'm doing for [zxing-cpp](https://github.com/glassechidna/zxing-cpp) and [zxing_cpp.rb](https://github.com/glassechidna/zxing_cpp.rb). For the most part, the former is simply included as a sub-module of the latter.

> The hope is obviously that decoupling ruby-vips from libvips will reduce maintenance.

That's certainly a reasonable goal. My only issue is that the two are naturally _somewhat_ tightly coupled. If the functionality offered by `ruby-vips` and `libvips` was essentially set in stone, then `ruby-vips` would be able to consume the exact same `libvips` bindings for a very long time - so this would be _less_ of an issue. However, the coupling becomes a bit more obvious as both libraries are under a fairly heavy development - which is great by the way, thank you so much for all your hard work!

> I've been deploying with Docker myself [...]

Thanks for sharing this. I'm also deploying to Docker in production, and setting up from source builds in my `Dockerfile` is not _too_ painful. Ideally, if possible, I'd rather dependencies be more explicitly tracked, and automatically resolved by my dependency manager.

**Comment by Benjamin-Dobell** [comment]
Seems as this seems to be a matter of personal preference, I'll happily close this issue as I don't believe there's likely to be anything actionable; no sense in polluting your issue tracker.

Again, thanks for your work, and your time responding to me!

**Comment by jcupitt** [comment]
I'm hoping that ruby-vips is pretty settled now. The plan anyway is to hardly touch it, and just add features to libvips in a backwards-compatible way. It will need to have the doc comments regenerated I guess.

**Comment by jcupitt** [comment]
I like your zxing gem. 

My feeling is that libvips has too many dependencies (up to about 40 I think) to be built in a simple way as part of `gem install`. It feels more like a platform library to me.

**Comment by Nakilon** [comment]
> polluting your issue tracker

This is exactly what issue trackers are for in my opinion.

I remember I had a hard time two years ago at my job to install it and I had to ask help of another guy. Later I was installing it on my own server and again I had to start tickets here IIRC. Of course for more experienced in Linux it would not be an issue but for a macOS guy like me - a user of bundler and brew -- this gem is not trivial, so I agree that it would be excellent if the next time I need it it would be some improvement like that I'd install it in one step without googling error messages ..D

But I also agree that including such long compilation into a gem would be too heavy.

How is it going in imagemagick? AFAIK it is splitted in two halves too, and even on macOS it has issues after you upgrade anything.
